# Actividad 1: Pipeline de procesamiento de datos con HDFS y Spark

## Actividad del Proyecto (actividad grupal)

Esta actividad está asociada al Proyecto transversal del título y para su desarrollo, tendrás que utilizar obligatoriamente el siguiente recurso:
* Dataset flights del Catálogo de Datos del proyecto

## Escribe aquí los nombres de los integrantes del grupo: Daniel Jiménez Alonso, Pedro Durán Porras, Juan Carlos Regueiro García, Juan Miguel Moreno Valero

## Recuerda borrar siempre las líneas que dicen `raise NotImplementedError`

Lee con detenimiento cada ejercicio. Las variables utilizadas para almacenar las soluciones, al igual que las nuevas columnas creadas, deben llamarse **exactamente** como indica el ejercicio, o de lo contrario los tests fallarán y el ejercicio no puntuará. Debe reemplazarse el valor `None` al que están inicializadas por el código necesario para resolver el ejercicio.

## Leemos el fichero flights.csv que hemos subido a Google Cloud Storage

Aunque para las capturas de pantalla se pide subir el fichero a HDFS, el resto de la actividad puede hacerse leyendo el mismo fichero que hemos subido al bucket de Google Cloud Storage.

Indicamos que contiene encabezados (nombres de columnas) y que intente inferir el esquema, aunque después comprobaremos si lo
ha inferido correctamente o no.

In [1]:
ruta_gcs = "/daniel_jimenez_alonso/flights.csv" # Reemplaza esto por la ruta correcta del fichero flights.csv en tu bucket de Google Cloud Storage
flightsDF = "gs://danieljimenezalonsounir_actividad1/actividad1/flights.csv"

# Descomentar estas líneas
flightsDF = spark.read\
            .option("header", "true")\
            .option("inferSchema", "true")\
            .csv(ruta_gcs)

# YOUR CODE HERE
#raise NotImplementedError

Imprimimos el esquema para comprobar qué tipo de dato ha inferido en cada columna

In [2]:
flightsDF.printSchema()

root
 |-- year: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- day: integer (nullable = true)
 |-- dep_time: string (nullable = true)
 |-- dep_delay: string (nullable = true)
 |-- arr_time: string (nullable = true)
 |-- arr_delay: string (nullable = true)
 |-- carrier: string (nullable = true)
 |-- tailnum: string (nullable = true)
 |-- flight: integer (nullable = true)
 |-- origin: string (nullable = true)
 |-- dest: string (nullable = true)
 |-- air_time: string (nullable = true)
 |-- distance: integer (nullable = true)
 |-- hour: string (nullable = true)
 |-- minute: string (nullable = true)



Mostramos el número de filas que tiene el DataFrame para hacernos una idea de su tamaño:

In [3]:
flightsDF.count()

162049

Vemos que tenemos 162049 filas. Si imprimimos por pantalla las 5 primeras filas, veremos qué tipos parecen tener y en qué columnas no coincide el tipo que podríamos esperar con el tipo que ha inferido Spark.

In [4]:
flightsDF.show(5)

+----+-----+---+--------+---------+--------+---------+-------+-------+------+------+----+--------+--------+----+------+
|year|month|day|dep_time|dep_delay|arr_time|arr_delay|carrier|tailnum|flight|origin|dest|air_time|distance|hour|minute|
+----+-----+---+--------+---------+--------+---------+-------+-------+------+------+----+--------+--------+----+------+
|2014|    1|  1|       1|       96|     235|       70|     AS| N508AS|   145|   PDX| ANC|     194|    1542|   0|     1|
|2014|    1|  1|       4|       -6|     738|      -23|     US| N195UW|  1830|   SEA| CLT|     252|    2279|   0|     4|
|2014|    1|  1|       8|       13|     548|       -4|     UA| N37422|  1609|   PDX| IAH|     201|    1825|   0|     8|
|2014|    1|  1|      28|       -2|     800|      -23|     US| N547UW|   466|   PDX| CLT|     251|    2282|   0|    28|
|2014|    1|  1|      34|       44|     325|       43|     AS| N762AS|   121|   SEA| ANC|     201|    1448|   0|    34|
+----+-----+---+--------+---------+-----

La causa del problema es que en muchas columnas existe un valor faltante llamado "NA". Spark no reconoce ese valor como
*no disponible* ni nada similar, sino que lo considera como un string de texto normal, y por tanto, asigna a toda la columna
el tipo de dato string (cadena de caracteres). Concretamente, las siguientes columnas deberían ser de tipo entero pero Spark
las muestra como string:
<ul>
 <li>dep_time: string (nullable = true)
 <li>dep_delay: string (nullable = true)
 <li>arr_time: string (nullable = true)
 <li>arr_delay: string (nullable = true)
 <li>air_time: string (nullable = true)
 <li>hour: string (nullable = true)
 <li>minute: string (nullable = true)    
</ul>


Vamos a averiguar cuántas filas tienen el valor "NA" (como string) en la columna dep_time:

In [5]:
from pyspark.sql import functions as F
cuantos_NA = flightsDF\
                .where(F.col("dep_time") == "NA")\
                .count()
cuantos_NA

857

Por tanto, hay 857 filas que no tienen un dato válido en esa columna. Hay distintas maneras de trabajar con los valores faltantes, como por ejemplo imputarlos (reemplazarlos por un valor generado por nosotros según cierta lógica, por ejemplo la media de esa columna, etc). Lo más sencillo es quitar toda la fila, aunque esto depende de si nos lo podemos permitir en base
a la cantidad de datos que tenemos. En nuestro caso, como tenemos un número considerable de filas, vamos a quitar todas las filas donde hay un NA en cualquiera de las columnas.

In [6]:
columnas_limpiar = ["dep_time", "dep_delay", "arr_time", "arr_delay", "air_time", "hour", "minute"]

flightsLimpiado = flightsDF
for nombreColumna in columnas_limpiar:  # para cada columna, nos quedamos con las filas que no tienen NA en esa columna
    flightsLimpiado = flightsLimpiado.where(F.col(nombreColumna) != "NA")

flightsLimpiado.cache()

DataFrame[year: int, month: int, day: int, dep_time: string, dep_delay: string, arr_time: string, arr_delay: string, carrier: string, tailnum: string, flight: int, origin: string, dest: string, air_time: string, distance: int, hour: string, minute: string]

Si ahora mostramos el número de filas que tiene el DataFrame `flightsLimpiado` tras eliminar todas esas filas, vemos que ha disminuido ligeramente
pero sigue siendo un número considerable como para realizar analítica y sacar conclusiones sobre estos datos

In [7]:
flightsLimpiado.count()

160748

Una vez que hemos eliminado los NA, vamos a convertir a tipo entero cada una de esas columnas que eran de tipo string. 
Ahora no debe haber problema ya que todas las cadenas de texto contienen dentro un número que puede ser convertido de texto a número. Vamos también a convertir la columna `arr_delay` de tipo entero a número real, necesario para los pasos posteriores donde ajustaremos un modelo predictivo.

In [8]:
from pyspark.sql.types import IntegerType, DoubleType

flightsConvertido = flightsLimpiado

for c in columnas_limpiar:
    # método que crea una columna o reemplaza una existente
    flightsConvertido = flightsConvertido.withColumn(c, F.col(c).cast(IntegerType())) 

flightsConvertido = flightsConvertido.withColumn("arr_delay", F.col("arr_delay").cast(DoubleType()))
flightsConvertido.cache()

DataFrame[year: int, month: int, day: int, dep_time: int, dep_delay: int, arr_time: int, arr_delay: double, carrier: string, tailnum: string, flight: int, origin: string, dest: string, air_time: int, distance: int, hour: int, minute: int]

In [9]:
flightsConvertido.printSchema()

root
 |-- year: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- day: integer (nullable = true)
 |-- dep_time: integer (nullable = true)
 |-- dep_delay: integer (nullable = true)
 |-- arr_time: integer (nullable = true)
 |-- arr_delay: double (nullable = true)
 |-- carrier: string (nullable = true)
 |-- tailnum: string (nullable = true)
 |-- flight: integer (nullable = true)
 |-- origin: string (nullable = true)
 |-- dest: string (nullable = true)
 |-- air_time: integer (nullable = true)
 |-- distance: integer (nullable = true)
 |-- hour: integer (nullable = true)
 |-- minute: integer (nullable = true)



Vamos a volver a mostrar las 5 primeras filas del DataFrame limpio. Aparentemente son iguales a las que ya teníamos, pero ahora
Spark sí está tratando como enteros las columnas que deberían serlo, y si queremos podemos hacer operaciones aritméticas
con ellas.

In [10]:
flightsConvertido.show(5)

+----+-----+---+--------+---------+--------+---------+-------+-------+------+------+----+--------+--------+----+------+
|year|month|day|dep_time|dep_delay|arr_time|arr_delay|carrier|tailnum|flight|origin|dest|air_time|distance|hour|minute|
+----+-----+---+--------+---------+--------+---------+-------+-------+------+------+----+--------+--------+----+------+
|2014|    1|  1|       1|       96|     235|     70.0|     AS| N508AS|   145|   PDX| ANC|     194|    1542|   0|     1|
|2014|    1|  1|       4|       -6|     738|    -23.0|     US| N195UW|  1830|   SEA| CLT|     252|    2279|   0|     4|
|2014|    1|  1|       8|       13|     548|     -4.0|     UA| N37422|  1609|   PDX| IAH|     201|    1825|   0|     8|
|2014|    1|  1|      28|       -2|     800|    -23.0|     US| N547UW|   466|   PDX| CLT|     251|    2282|   0|    28|
|2014|    1|  1|      34|       44|     325|     43.0|     AS| N762AS|   121|   SEA| ANC|     201|    1448|   0|    34|
+----+-----+---+--------+---------+-----

In [11]:
flightsConvertido.groupBy("carrier").count().show()

+-------+-----+
|carrier|count|
+-------+-----+
|     UA|16452|
|     AA| 7474|
|     B6| 3493|
|     DL|16637|
|     OO|18368|
|     F9| 2683|
|     US| 5876|
|     HA| 1092|
|     AS|62189|
|     VX| 3266|
|     WN|23218|
+-------+-----+



### Ejercicio 1 (3 puntos) 
### (sólo puntúa 3 puntos si pasa la celda de autoevaluación sin errores, o 0 puntos si hay algún error. No hay puntaje intermedio)

Partiendo del DataFrame `flightsConvertido` que ya tiene los tipos correctos en las columnas, se pide **encadenar transformaciones sin crear ninguna variable intermedia**, de manera que se vayan creando las siguientes columnas:

* Una nueva columna llamada `flight_date` con la fecha del vuelo como objeto fecha. Utiliza para ello la función `F.make_date` aplicada a los argumentos (columnas) `year`, `month` y `day` tal como indica la documentación de `make_date` disponible [aquí](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.make_date.html).
* Encadenado con la transformación anterior, crear una nueva columna llamada `day_of_week` con el día de la semana correspondiente al vuelo, utilizando la función `F.dayofweek` cuya documentación puedes ver [aquí](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.dayofweek.html). Debes pasarle como argumento la columna `flight_date` creada en el apartado anterior.
* Encadenado con la transformación anterior, agrupar para obtener **tantas filas como aeropuertos de destino existan, y tantas columnas como días de la semana** hay en el DF original. 
  * Cada celda del DF resultante debe contener el *conteo* del número de vuelos que existen para ese destino en ese día de la semana.
  * El DF resultante de la agrupación seguida de pivot tendrá 8 columnas, que serán el destino, y los 7 días de la semana, llamados del 1 al 7.
* Encadenado con la transformación anterior, rellenar los valores nulos con un 0, puesto que todos los valores representan conteos, y los nulos han sido provocados porque ciertas combinaciones de (origen, destino, día de la semana) no existen en los datos (no hay vuelos para ciertas rutas en ciertos días de la semana). Por eso, tiene sentido sustituir los nulos por 0, que es el conteo para esa combinación no encontrada.
* Encadenado con la transformación anterior, renombrar las columnas de los días de la semana para que pasen a llamarse Domingo, Lunes, ..., Sabado (sin tildes). Puedes hacerlo encadenando `withColumnRenamed` varias veces, o bien con `selectExpr` utilizando `as` en cada argumento de tipo string. Consulta la documentación de `selectExpr` [aquí](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.selectExpr.html).
  * Si utilizas selectExpr, entonces debes referirte a las columnas como ``"`1`"`` (con acento grave) y no como `"1"` ya que, al tener por nombre un número entero, hay que referirse a ellas de ese modo. Si utilizas withColumnRenamed, no es necesario y puedes referirte a ellas como `"1"`.
  * Si lo deseas, puedes recategorizar la columna day_of_week, justo después de crearla, utilizando para ello otra llamada a `withColumn` para reemplazar dicha columna, usando `F.when` dentro de ella. De este modo, al llegar al groupBy(...).pivot(...) ya tendrá los nombres de columna correctos y no será necesario renombrar columnas.
* Encadenado con la transformación anterior, añadir una nueva columna llamada `n_vuelos` que contenga, en cada fila, la suma de las columnas de los 7 días de la semana. Esto representa el número de vuelos totales para ese destino, independientemente del día del vuelo.
* Encadenado con la transformación anterior, generar un nuevo DF que esté ordenado por la columna `n_vuelos` *descendentemente*.
* El DF resultante de todas estas transformaciones, que deben encadenarse sin utilizar ninguna otra variable, debe quedar guardado en la variable `conteos_df`.
* No está permitido utilizar construcciones de la forma df = df.algo(). **Sólo puede utilizarse una vez el operador de asignación (el signo `=`).**

In [12]:
from pyspark.sql import functions as F
# Reemplaza None por el código necesario para calcular sus valores correctos
conteos_df = flightsConvertido\
            .withColumn('flight_date', F.make_date("year", "month", "day"))\
            .withColumn('day_of_week', F.dayofweek('flight_date'))\
            .groupBy('dest')\
            .pivot('day_of_week', [1,2,3,4,5,6,7])\
            .agg(F.count('*').alias('count'))\
            .selectExpr("dest", "`1` as Domingo","`2` as Lunes","`3` as Martes","`4` as Miercoles","`5` as Jueves", "`6` as Viernes","`7` as Sabado")\
            .fillna(0)\
            .withColumn('n_vuelos', sum(F.col(day) for day in ["Domingo", "Lunes", "Martes", "Miercoles", "Jueves", "Viernes", "Sabado"]))\
            .orderBy(F.col('n_vuelos').desc())\
            
#Comprobacion tabla
conteos_df.show(5)


# YOUR CODE HERE
#raise NotImplementedError

+----+-------+-----+------+---------+------+-------+------+--------+
|dest|Domingo|Lunes|Martes|Miercoles|Jueves|Viernes|Sabado|n_vuelos|
+----+-------+-----+------+---------+------+-------+------+--------+
| SFO|   1723| 1922|  1840|     1863|  1917|   1910|  1471|   12646|
| LAX|   1473| 1550|  1502|     1541|  1550|   1545|  1248|   10409|
| DEN|   1338| 1423|  1348|     1385|  1411|   1376|  1152|    9433|
| PHX|   1211| 1259|  1203|     1272|  1260|   1254|  1142|    8601|
| LAS|   1128| 1205|  1166|     1176|  1200|   1194|  1113|    8182|
+----+-------+-----+------+---------+------+-------+------+--------+
only showing top 5 rows



In [13]:
from pyspark.sql import functions as F
c = conteos_df.columns
assert(len(c) == 9)  # debe tener 9 columnas que son dest, los días de la semana, y n_vuelos (no necesariamente en ese orden)
assert(all([x in c for x in ["dest", "n_vuelos", "Domingo", "Lunes", "Martes", "Miercoles", "Jueves", "Viernes", "Sabado"]]))  # Comprobar nombres de columnas
cnt_list = conteos_df.take(100)
assert(cnt_list[0].dest == "SFO" and cnt_list[0].Viernes == 1910)
assert(cnt_list[1].dest == "LAX" and cnt_list[1].Lunes == 1550)
assert(cnt_list[2].dest == "DEN" and cnt_list[2].n_vuelos == 9433)
assert(cnt_list[-1].Domingo == 0 and cnt_list[-1].n_vuelos == 2)  # la ruta con menos vuelos de todos. Nos aseguramos de que se hayan rellenado los nulos con 0.

### Ejercicio 2 (3 puntos)
### (sólo puntúa 3 puntos si pasa la celda de autoevaluación sin errores, o 0 puntos si hay algún error. No hay puntaje intermedio)

Partiendo de nuevo de `flightsConvertido`, se pide crear una función `retraso_medio_periodo`, que reciba como argumento un DF en el podemos asumir que existen las columnas `dep_time` y `arr_delay` y haga lo siguiente:

* Añada una nueva columna `periodo_dia` para recategorizar `dep_time` con los siguientes valores sin tildes (puedes hacerlo con withColumn y F.when):
  * `"maniana"` cuando la hora de salida del vuelo esté entre las 7 de la mañana no incluida (expresado como 700 en dep_time) y las 12 de la mañana incluida (expresado como 1200)
  * `"mediodia"` cuando la hora de salida del vuelo esté entre las 12 de la mañana no incluida (expresado como 1200) y las 5 de la tarde incluida (expresado como 1700)
  * `"tarde"` cuando la hora de salida del vuelo esté entre las 5 de la tarde no incluida (expresado como 1700) y las 9 de la noche incluida (expresado como 2100)
  * `"noche"` cuando la hora de salida del vuelo esté entre las 9 de la noche no incluida (expresado como 2100) y las 7 de la mañana incluida (expresado como 700)
  * PISTAS:
    * Recuerda que las condiciones compuestas con columnas booleanas en Spark requieren obligatoriamente utilizar paréntesis envolviendo a cada condición simple.
    * Recuerda también utilizar F.lit(...) para indicar el valor (constante) de la nueva columna para cada una de estas condiciones.
    * Cuidado con la noche: la condición es que la hora sea mayor que 2100 **o bien** (condición `|`) menor o igual que 700. En el resto de condiciones, necesitamos `&`.

* Encadenado con la transformación anterior, sólo para los vuelos que llegan con **retraso estrictamente positivo (>0)**, calcular el retraso medio a la llegada de dichos vuelos **para cada aeropuerto de destino y cada hora del día**, desplegando las franjas horarias como columnas independientes. 
  * El DF calculado debe tener tantas columnas como franjas horarias más la columna del aeropuerto de destino (es decir, 5 columnas en total)

* Encadenado con la transformación anterior, tras haber hecho la agrupación y agregación, añadir una columna constante (con `F.lit`) llamada `alumno` que contenga el primer apellido de cada miembro del grupo (ejemplo: si el grupo lo forman Pablo García, Francisco Pérez y Antonio Martín, el valor de la columna debe ser `García_Pérez_Martín`).
* Encadenado con la transformación anterior, ordenar alfabéticamente el DF resultante por aeropuerto de destino, ascendentemente. El resultado de la ordenación será un nuevo DF que debe ser devuelto como resultado de la función (recuerda que es imposible modificar el DF original).

In [14]:
from pyspark.sql import DataFrame

def retraso_medio_periodo(df: DataFrame) -> DataFrame:  # El argumento que recibe la función es un DataFrame de Spark, y el resultado devuelto también
    df = df.withColumn("periodo_dia", 
            F.when((F.col("dep_time") > 700) & (F.col("dep_time") <= 1200), F.lit("maniana"))
             .when((F.col("dep_time") > 1200) & (F.col("dep_time") <= 1700), F.lit("mediodia"))
             .when((F.col("dep_time") > 1700) & (F.col("dep_time") <= 2100), F.lit("tarde"))
             .when((F.col("dep_time") > 2100) | (F.col("dep_time") <= 700), F.lit("noche")))\
           .filter(F.col("arr_delay") > 0)\
           .groupBy("dest")\
           .pivot("periodo_dia", ["maniana", "mediodia", "tarde", "noche"])\
           .agg(F.avg("arr_delay").alias("retraso_medio"))\
           .fillna(0)\
           .withColumn("alumno", F.lit("Jimenez_Regueiro_Moreno_Duran"))\
           .orderBy(F.col('dest').asc())    
    return df  # Reemplaza None por la variable que quieras devolver

#COMPROBACIONES
retraso = retraso_medio_periodo(flightsConvertido)
retraso.show(3)




# YOUR CODE HERE
#raise NotImplementedError

+----+------------------+------------------+-----------------+------------------+--------------------+
|dest|           maniana|          mediodia|            tarde|             noche|              alumno|
+----+------------------+------------------+-----------------+------------------+--------------------+
| ABQ|12.912280701754385|23.872340425531913|             99.4| 8.333333333333334|Jimenez_Regueiro_...|
| ANC|15.314893617021276|16.876840696117803|18.20250521920668|20.703105590062112|Jimenez_Regueiro_...|
| ATL|19.243291592128802|             28.75|            375.0|18.298200514138816|Jimenez_Regueiro_...|
+----+------------------+------------------+-----------------+------------------+--------------------+
only showing top 3 rows



In [15]:
retraso_medio_df = retraso_medio_periodo(flightsConvertido)
lista = retraso_medio_df.take(3)
c = retraso_medio_df.columns
assert(len(c) == 6)  #  el DF resultante debe tener 6 columnas
assert(all([x in c for x in ["dest", "maniana", "mediodia", "tarde", "noche", "alumno"]]))  # Comprobamos los nombres de columnas
assert(lista[0].dest == "ABQ" and round(lista[0].maniana, 2) == 12.91 and round(lista[0].noche, 2) == 8.33)
assert(lista[1].dest == "ANC" and round(lista[1].mediodia, 2) == 16.88 and round(lista[1].noche, 2) == 20.70)
assert(lista[2].dest == "ATL" and round(lista[2].tarde, 2) == 375.0 and round(lista[2].noche, 2) == 18.30)

Ahora invocamos a nuestra función `retrasoMedio` pasándole como argumento `flightsConvertido`. ¿Cuáles son los tres aeropuertos con mayor retraso medio? ¿Cuáles son sus retrasos medios en minutos?

### Ejercicio 3

Ajustar un modelo de DecisionTree de Spark para predecir si un vuelo vendrá o no con retraso (problema de clasificación binaria), utilizando como variables predictoras el mes, el día del mes, la hora de partida `dep_time`, la hora de llegada `arr_time`, el tipo de avión (`carrier`), la distancia y el tiempo que permanece en el aire. Para ello, sigue los siguientes pasos.

Notemos que en estos datos hay variables numéricas y variables categóricas que ahora mismo están tipadas como numéricas, como por ejemplo el mes del año (`month`), que es en realidad categórica. Debemos indicar a Spark cuáles son categóricas e indexarlas. Para ello se pide: 

* **(1 punto)** Crear un `StringIndexer` llamado `indexerMonth` y otro llamado `indexerCarrier` sobre las variables categóricas `month` y `carrier` (tipo de avión). El nombre de las columnas indexadas que se crearán debe ser, respectivamente, `monthIndexed` y `carrierIndexed`. Asegúrate de que, cuando necesite codificar una categoría que no existía cuando se entrenó, esta pieza no lance un error sino que le asigne el primer valor que esté libre (argumento `handleInvalid`).

In [16]:
# Incluye aquí los imports que necesites
from pyspark.ml.feature import StringIndexer

indexerMonth = StringIndexer(inputCol="month", outputCol="monthIndexed",handleInvalid="keep")
indexerCarrier = StringIndexer(inputCol="carrier", outputCol="carrierIndexed",handleInvalid="keep")

# YOUR CODE HERE
#raise NotImplementedError

In [17]:
assert(isinstance(indexerMonth, StringIndexer))
assert(isinstance(indexerCarrier, StringIndexer))
assert(indexerMonth.getInputCol() == "month" and indexerMonth.getOutputCol() == "monthIndexed" and indexerMonth.getHandleInvalid() == "keep")
assert(indexerCarrier.getInputCol() == "carrier" and indexerCarrier.getOutputCol() == "carrierIndexed" and indexerCarrier.getHandleInvalid() == "keep")

Recordemos también que Spark requiere que todas las variables estén en una única columna de tipo vector, por lo que después de indexar estas dos variables, tendremos que fusionar en una columna de tipo vector todas ellas, utilizando un `VectorAssembler`. Se pide:

* **(1 punto)** Crear en una variable llamada `vectorAssembler` un `VectorAssembler` que reciba como entrada una lista de todas las variables de entrada (y que no debe incluir `arr_delay`) que serán las que formarán parte del modelo. Crear primero esta lista de variables (lista de strings) en la variable `columnas_ensamblar` y pasar dicha variable como argumento al crear el `VectorAssembler`. Como es lógico, en el caso de las columnas `month` y `carrier`, no usaremos las variables originales sino las indexadas en el apartado anterior. La columna de tipo vector creada con las características ensambladas debe llamarse `features`.

In [18]:
# Incluye aquí los imports que necesites
# import ...........

from pyspark.ml.feature import VectorAssembler

columnas_ensamblar = ["monthIndexed","day", "dep_time", "arr_time", "carrierIndexed", "air_time", "distance"]
print("Longitud Columnas Ensamblar: " ,len(columnas_ensamblar))
outputCol = "features"
vectorAssembler = VectorAssembler(inputCols = columnas_ensamblar, outputCol=outputCol)

# YOUR CODE HERE
#raise NotImplementedError

Longitud Columnas Ensamblar:  7


In [19]:
assert(isinstance(vectorAssembler, VectorAssembler))
assert(vectorAssembler.getOutputCol() == "features")
input_cols = vectorAssembler.getInputCols()
assert(len(input_cols) == 7)
assert("arr_delay" not in input_cols)

Finalmente, vemos que la columna `arr_delay` es continua, y no binaria como requiere un problema de clasificación con dos clases. Vamos a convertirla en binaria. Para ello se pide:

* **(0.5 puntos)** Utilizar un binarizador de Spark, fijando a 15 el umbral, y guardarlo en la variable `delayBinarizer`. Consideramos retrasado un vuelo que ha llegado con más de 15 minutos de retraso, y no retrasado en caso contrario. La nueva columna creada con la variable binaria debe llamarse `arr_delay_binary` y debe ser interpretada como la columna target para nuestro algoritmo. Por ese motivo, esta columna **no** se incluyó en el apartado anterior entre las columnas que se ensamblan para formar las features.

In [20]:
# Incluye aquí los imports que necesites y que no hayas incluido ya en alguna celda anterior
# import .........

from pyspark.ml.feature import Binarizer

delayBinarizer = Binarizer(inputCol="arr_delay", outputCol="arr_delay_binary", threshold=15.0)

# YOUR CODE HERE
#raise NotImplementedError

In [21]:
assert(isinstance(delayBinarizer, Binarizer))
assert(delayBinarizer.getThreshold() == 15)
assert(delayBinarizer.getInputCol() == "arr_delay")
assert(delayBinarizer.getOutputCol() == "arr_delay_binary")

**(0.5 puntos)** Por último, crearemos el modelo de clasificación.

* Crear en una variable `decisionTree` un árbol de clasificación de Spark (`DecisionTreeClassifier` del paquete `pyspark.ml.classification`)
* Indicar como columna de entrada la nueva columna creada por el `VectorAssembler` creado en un apartado anterior.
* Indicar como columna objetivo (target) la nueva columna creada por el `Binarizer` del apartado anterior.

In [22]:
# Incluye aquí los imports que necesites y que no hayas incluido ya en alguna celda anterior
# import .........

from pyspark.ml.classification import DecisionTreeClassifier

decisionTree = DecisionTreeClassifier(featuresCol="features", labelCol="arr_delay_binary")

# YOUR CODE HERE
#raise NotImplementedError

In [23]:
assert(isinstance(decisionTree, DecisionTreeClassifier))
assert(decisionTree.getFeaturesCol() == "features")
assert(decisionTree.getLabelCol() == "arr_delay_binary")

**(1 punto)** Ahora vamos a encapsular todas las fases en un sólo pipeline y procederemos a entrenarlo. Se pide:

* Crear en una variable llamada `pipeline` un objeto `Pipeline` de Spark con las etapas anteriores en el orden adecuado para poder entrenar un modelo. 

* Entrenarlo invocando sobre ella al método `fit` y guardar el pipeline entrenado devuelto por dicho método en una variable llamada `pipelineModel`. 

* Aplicar el pipeline entrenado para transformar (predecir) el DataFrame `flightsConvertido`, guardando las predicciones devueltas en la variable `flightsPredictions` que será un DataFrame. Nótese que estamos prediciendo los propios datos de entrenamiento y que, por simplicidad, no habíamos hecho (aunque habría sido lo correcto) ninguna división de nuestros datos originales en subconjuntos distintos de entrenamiento y test antes de entrenar.

In [24]:
# Incluye aquí los imports que necesites y que no hayas incluido ya en alguna celda anterior
# import .........

from pyspark.ml import Pipeline
from pyspark.ml import PipelineModel

pipeline = Pipeline(stages=[indexerMonth,indexerCarrier, vectorAssembler, delayBinarizer, decisionTree])
pipelineModel = pipeline.fit(flightsConvertido)
flightsPredictions = pipelineModel.transform(flightsConvertido)

# YOUR CODE HERE
#raise NotImplementedError

In [25]:
from pyspark.ml import PipelineModel
assert(isinstance(pipeline, Pipeline))
assert(len(pipeline.getStages()) == 5)
assert(isinstance(pipelineModel, PipelineModel))
assert("probability" in flightsPredictions.columns)
assert("prediction" in flightsPredictions.columns)
assert("rawPrediction" in flightsPredictions.columns)

Vamos a mostrar la matriz de confusión (este apartado no es evaluable). Agrupamos por la variable que tiene la clase verdadera y la que tiene la clase predicha, para ver en cuántos casos coinciden y en cuántos difieren.

In [26]:
flightsPredictions.groupBy("arr_delay_binary", "prediction").count().show()

+----------------+----------+------+
|arr_delay_binary|prediction| count|
+----------------+----------+------+
|             1.0|       1.0|   505|
|             0.0|       1.0|    73|
|             1.0|       0.0| 23744|
|             0.0|       0.0|136426|
+----------------+----------+------+

